In [10]:
# Q1: POS-based Extractor
import wikipedia
import spacy

nlp = spacy.load("en_core_web_sm")
page_title_q1 = "Brad Pitt"
page_content_q1 = wikipedia.page(page_title_q1, auto_suggest=False).content
doc_q1 = nlp(page_content_q1)

proper_noun_spans = []
i = 0
while i < len(doc_q1):
    if doc_q1[i].pos_ == "PROPN":
        start = i
        while i < len(doc_q1) and doc_q1[i].pos_ == "PROPN":
            i += 1
        end = i
        span_text = doc_q1[start:end].text
        proper_noun_spans.append((span_text, start, end))
    else:
        i += 1

triplets_q1 = []
for idx in range(len(proper_noun_spans) - 1):
    subj_span, subj_start, subj_end = proper_noun_spans[idx]
    obj_span, obj_start, obj_end = proper_noun_spans[idx + 1]
    in_between_tokens = doc_q1[subj_end:obj_start]

    if any(t.pos_ == "PUNCT" for t in in_between_tokens):
        continue

    if not any(t.pos_ == "VERB" for t in in_between_tokens):
        continue

    relation_tokens = [t.text for t in in_between_tokens if t.pos_ in ["VERB", "ADP"]]
    relation_text = " ".join(relation_tokens).strip()

    if relation_text:
        triplets_q1.append((subj_span, relation_text, obj_span))

print("Q1 Triplets (POS-based)")
for subj, rel, obj_ in triplets_q1:
    print(f"Subject: {subj}\tRelation: {rel}\tObject: {obj_}")


=== Q1 Triplets (POS-based) ===
Subject: Pitt	Relation: appeared on	Object: Forbes
Subject: Pitt	Relation: emerged as taking on leading in as	Object: River Runs
Subject: Pitt	Relation: found starring in	Object: Steven Soderbergh
Subject: Best Supporting Actor	Relation: for playing in	Object: Quentin Tarantino
Subject: Pitt	Relation: co - founded	Object: Plan B Entertainment
Subject: Pitt	Relation: named	Object: People
Subject: William Bradley Pitt	Relation: born on	Object: December
Subject: Pitt	Relation: described	Object: Springfield
Subject: Pitt	Relation: attended	Object: Kickapoo High School
Subject: Pitt	Relation: enrolled in	Object: University
Subject: Pitt	Relation: left moved to	Object: Los Angeles
Subject: Pitt	Relation: took from acting	Object: Roy London
Subject: Pitt	Relation: had on	Object: CBS
Subject: Pitt	Relation: made on	Object: Fox
Subject: American	Relation: taken by to	Object: Adriatic
Subject: Pitt	Relation: cast as	Object: Billy Canton
Subject: Glory Days	Relatio

Q1 Triplets (POS-based)
Subject: Pitt	Relation: appeared on	Object: Forbes
Subject: Pitt	Relation: emerged as taking on leading in as	Object: River Runs
Subject: Pitt	Relation: found starring in	Object: Steven Soderbergh
Subject: Best Supporting Actor	Relation: for playing in	Object: Quentin Tarantino
Subject: Pitt	Relation: co - founded	Object: Plan B Entertainment
Subject: Pitt	Relation: named	Object: People
Subject: William Bradley Pitt	Relation: born on	Object: December
Subject: Pitt	Relation: described	Object: Springfield
Subject: Pitt	Relation: attended	Object: Kickapoo High School
Subject: Pitt	Relation: enrolled in	Object: University
Subject: Pitt	Relation: left moved to	Object: Los Angeles
Subject: Pitt	Relation: took from acting	Object: Roy London
Subject: Pitt	Relation: had on	Object: CBS
Subject: Pitt	Relation: made on	Object: Fox
Subject: American	Relation: taken by to	Object: Adriatic
Subject: Pitt	Relation: cast as	Object: Billy Canton
Subject: Glory Days	Relation: took supporting in	Object: HBO
Subject: Pitt	Relation: portrayed	Object: Joe Maloney
Subject: Pitt	Relation: attracted in supporting in	Object: Ridley Scott
Subject: Davis	Relation: cited as defined	Object: Pitt
Subject: Pitt	Relation: starred in	Object: Johnny Suede
Subject: Pitt	Relation: reunited with	Object: Juliette Lewis
Subject: Pitt	Relation: garnered for in hit	Object: True Romance
Subject: True Romance	Relation: as named	Object: Floyd
Subject: Pitt	Relation: portrayed	Object: Louis de Pointe du Lac
Subject: Pitt	Relation: starred in	Object: Legends
Subject: Pitt	Relation: received	Object: Golden Globe Award
Subject: Henry Thomas co	Relation: - starred as	Object: Pitt
Subject: Pitt	Relation: of acting works to	Object: heartthrob
Subject: Deseret News	Relation: predicted of	Object: Fall
Subject: Fall	Relation: solidify	Object: Pitt
Subject: Pitt	Relation: starred alongside	Object: Morgan Freeman
Subject: Pitt	Relation: played	Object: Jeffrey Goines
Subject: New York Times	Relation: called	Object: Twelve Monkeys
Subject: Best Supporting Actor	Relation: for received	Object: Academy Award
Subject: Charleston Gazette	Relation: opined favored	Object: Pitt
Subject: Pitt	Relation: had in	Object: Meet Joe Black
Subject: Pitt	Relation: portrayed	Object: Tyler Durden
Subject: Fight Club	Relation: premiered at	Object: Venice International Film Festival
Subject: Variety	Relation: remarked	Object: Pitt
Subject: Pitt	Relation: cast as	Object: Irish Traveller
Subject: San Francisco Chronicle	Relation: said	Object: Pitt
Subject: Pitt	Relation: starred opposite	Object: Julia Roberts
Subject: Pitt	Relation: made in of	Object: Friends
Subject: Pitt	Relation: played	Object: Rusty Ryan
Subject: Pitt	Relation: had in	Object: George Clooney
Subject: DreamWorks	Relation: animated	Object: Sinbad
Subject: Seven Seas	Relation: playing	Object: Boomhauer
Subject: Washington Post	Relation: stated	Object: Pitt
Subject: Troy	Relation: produced by	Object: Plan B Entertainment
Subject: Clooney	Relation: described by	Object: CNN
Subject: Pitt	Relation: starred as	Object: John Smith
Subject: Pitt	Relation: starred	Object: Cate Blanchett
Subject: Cannes Film Festival	Relation: featured at	Object: Toronto International Film Festival
Subject: Babel	Relation: received	Object: Academy
Subject: Plan B Entertainment	Relation: produced	Object: Departed
Subject: Graham King	Relation: ruled for	Object: Oscar
Subject: Pitt	Relation: starred in	Object: Ocean
Subject: Andrew Dominik	Relation: produced by	Object: Pitt
Subject: Pitt	Relation: leading came in with	Object: Quentin Tarantino
Subject: Pitt	Relation: played	Object: Lieutenant Aldo Raine
Subject: Moneyball	Relation: received	Object: Academy Award
Subject: Academy Award	Relation: including	Object: Best Picture
Subject: Pitt	Relation: starred in	Object: World War Z
Subject: World War Z	Relation: grossed at	Object: box
Subject: Pitt	Relation: had supporting in	Object: Ridley Scott
Subject: World War II	Relation: directed written by	Object: David Ayer
Subject: Pitt	Relation: starred in	Object: Robert Zemeckis
Subject: Pitt	Relation: starred as	Object: Cliff Booth
Subject: Pitt	Relation: produced	Object: DiCaprio
Subject: Pitt	Relation: portrayed	Object: Dr. Anthony Fauci
Subject: Pitt	Relation: entered by creating with	Object: Damien Quintard
Subject: Pitt	Relation: starred in	Object: Bullet Train
Subject: Formula One	Relation: titled	Object: F1
Subject: Pitt	Relation: received including	Object: Academy Awards
Subject: Hollywood	Relation: =	Object: Pitt
Subject: Pitt	Relation: visited	Object: University
Subject: October	Relation: encourage vote in	Object: U.S.
Subject: Jolie	Relation: flew to	Object: Haiti
Subject: Jolie	Relation: donated in	Object: Chad
Subject: Sudan	Relation: dedicated to affected by in	Object: Darfur
Subject: Ninth Ward	Relation: following caused by	Object: Hurricane Katrina
Subject: Pitt	Relation: received in of from	Object: U.S. Green Building Council
Subject: Pitt	Relation: met with	Object: U.S. President Barack Obama
Subject: Pitt Foundation	Relation: donated to	Object: U.N.
Subject: U.N.	Relation: help	Object: Pakistanis
Subject: Pakistanis	Relation: displaced by fighting between	Object: Taliban
Subject: Borders	Relation: for help of	Object: Haiti
Subject: Pitt	Relation: said marry	Object: Jolie
Subject: Pitt	Relation: featured in of	Object: Dustin Lance Black
Subject: Pitt	Relation: reaffirmed for	Object: Obama
Subject: Obama	Relation: backing	Object: U.S.
Subject: Aniston	Relation: finalized by	Object: Los Angeles Superior Court
Subject: Pitt	Relation: said in	Object: February
Subject: Jolie	Relation: confirmed to with	Object: Pitt
Subject: Jolie	Relation: announced in	Object: April
Subject: Jolie	Relation: filed for from	Object: Pitt
Subject: Pitt	Relation: sued	Object: Jolie
Subject: Pitt	Relation: accompanied	Object: Jolie
Subject: Pitt	Relation: in of becoming of	Object: Jolie
Subject: California	Relation: granted	Object: Jolie
Subject: Jolie	Relation: gave to	Object: Shiloh
Subject: Pitt	Relation: adopted in	Object: United States
Subject: Jolie	Relation: filed for from	Object: Pitt
Subject: Jolie	Relation: had until	Object: May
Subject: John W. Ouderkirk	Relation: removed from due to over disclose with	Object: Pitt
Subject: Jolie	Relation: has	Object: Pitt
Subject: Family Services	Relation: investigated	Object: Pitt
Subject: Pitt	Relation: charged for by	Object: Los Angeles County Department
Subject: Jolie	Relation: filed against	Object: Pitt
Subject: Pitt	Relation: asked	Object: Jolie
Subject: Pitt	Relation: got began attending	Object: Alcoholics Anonymous
Subject: Interview	Relation: in talked with	Object: Legends
Subject: Meet Joe Black	Relation: recovering	Object: Anthony Hopkins
Subject: Pitt	Relation: said struggled for recognize due to	Object: prosopagnosia
Subject: Pitt	Relation: diagnosed with	Object: prosopagnosia
Subject: Nick Cave	Relation: artist	Object: Thomas Houseago
Subject: Pitt	Relation: appeared on	Object: Forbes
Subject: Jolie	Relation: spent in in	Object: Nice
Subject: Heineken	Relation: aired during	Object: Super Bowl
Subject: Brad Pitt	Relation: collected at	Object: New York Times
Subject: Brad Pitt	Relation: collected at	Object: The Guardian

In [11]:
# Q2: Dependency-based Extractor
import wikipedia
import spacy

nlp = spacy.load("en_core_web_sm")

page_title_q2 = "Brad Pitt"
page_content_q2 = wikipedia.page(page_title_q2, auto_suggest=False).content
doc_q2 = nlp(page_content_q2)

proper_noun_heads = []
for token in doc_q2:
    if token.pos_ == "PROPN" and token.dep_ != "compound":
        proper_noun_heads.append(token)

proper_noun_map = {}
for head in proper_noun_heads:
    pn_list = [head.text]
    for child in head.children:
        if child.dep_ == "compound" and child.pos_ == "PROPN":
            pn_list.insert(0, child.text)
    proper_noun_map[head] = " ".join(pn_list)

def condition_1(h1, h2):
    return (
        h1.head == h2.head
        and h1.dep_ == "nsubj"
        and h2.dep_ == "dobj"
    )

def condition_2(h1, h2):
    h = h1.head
    p = h2.head
    return (
        h1.dep_ == "nsubj"
        and h2.dep_ == "pobj"
        and p.dep_ == "prep"
        and p.head == h
    )

triplets_q2 = []
for i in range(len(proper_noun_heads)):
    for j in range(len(proper_noun_heads)):
        if i == j:
            continue
        h1 = proper_noun_heads[i]
        h2 = proper_noun_heads[j]

        if condition_1(h1, h2):
            verb = h1.head.text
            subj_text = proper_noun_map[h1]
            obj_text  = proper_noun_map[h2]
            triplets_q2.append((subj_text, verb, obj_text))
        elif condition_2(h1, h2):
            main_verb = h1.head.text
            prep = h2.head.text
            relation_text = main_verb + " " + prep
            subj_text = proper_noun_map[h1]
            obj_text  = proper_noun_map[h2]
            triplets_q2.append((subj_text, relation_text, obj_text))

print("Q2 Triplets (Dependency-based)")
for subj, rel, obj_ in triplets_q2:
    print(f"Subject: {subj}\tRelation: {rel}\tObject: {obj_}")



=== Q2 Triplets (Dependency-based) ===
Subject: Pitt	Relation: described	Object: Springfield
Subject: Pitt	Relation: attended	Object: High Kickapoo School
Subject: Pitt	Relation: enrolled in	Object: University
Subject: Pitt	Relation: took from	Object: Roy London
Subject: Pitt	Relation: had In	Object: November
Subject: Pitt	Relation: reunited with	Object: Juliette Lewis
Subject: Pitt	Relation: portrayed	Object: du Louis Lac
Subject: Pitt	Relation: starred in	Object: Legends
Subject: Pitt	Relation: starred alongside	Object: Morgan Freeman
Subject: Pitt	Relation: played	Object: Jeffrey Goines
Subject: Janet Maslin	Relation: called	Object: Twelve Monkeys
Subject: Pitt	Relation: starred opposite	Object: Harrison Ford
Subject: Pitt	Relation: starred as	Object: Rory Devany
Subject: Pitt	Relation: portrayed	Object: Tyler Durden
Subject: Fight Club	Relation: premiered at	Object: Film International Venice Festival
Subject: Pitt	Relation: starred opposite	Object: Julia Roberts
Subject: Pitt	Rela

Q2 Triplets (Dependency-based)
Subject: Pitt	Relation: described	Object: Springfield
Subject: Pitt	Relation: attended	Object: High Kickapoo School
Subject: Pitt	Relation: enrolled in	Object: University
Subject: Pitt	Relation: took from	Object: Roy London
Subject: Pitt	Relation: had In	Object: November
Subject: Pitt	Relation: reunited with	Object: Juliette Lewis
Subject: Pitt	Relation: portrayed	Object: du Louis Lac
Subject: Pitt	Relation: starred in	Object: Legends
Subject: Pitt	Relation: starred alongside	Object: Morgan Freeman
Subject: Pitt	Relation: played	Object: Jeffrey Goines
Subject: Janet Maslin	Relation: called	Object: Twelve Monkeys
Subject: Pitt	Relation: starred opposite	Object: Harrison Ford
Subject: Pitt	Relation: starred as	Object: Rory Devany
Subject: Pitt	Relation: portrayed	Object: Tyler Durden
Subject: Fight Club	Relation: premiered at	Object: Film International Venice Festival
Subject: Pitt	Relation: starred opposite	Object: Julia Roberts
Subject: Pitt	Relation: made On	Object: November
Subject: Pitt	Relation: played In	Object: December
Subject: Pitt	Relation: played	Object: Rusty Ryan
Subject: Pitt	Relation: played in	Object: Eleven
Subject: Pitt	Relation: starred as	Object: John Smith
Subject: Pitt	Relation: starred	Object: Cate Blanchett
Subject: B Entertainment	Relation: produced	Object: Departed
Subject: Pitt	Relation: starred in	Object: Thirteen
Subject: Pitt	Relation: played	Object: Aldo Lieutenant Raine
Subject: Moneyball	Relation: received for	Object: Pitt
Subject: Pitt	Relation: starred in	Object: War World Z
Subject: Pitt	Relation: starred By	Object: Sea
Subject: Pitt	Relation: starred as	Object: Cliff Booth
Subject: Pitt	Relation: starred opposite	Object: Leonardo DiCaprio
Subject: Pitt	Relation: portrayed	Object: Anthony Dr. Fauci
Subject: Pitt	Relation: portrayed on	Object: Night Saturday Live
Subject: Pitt	Relation: starred in	Object: Bullet Train
Subject: Pitt	Relation: visited in	Object: October
Subject: Pitt	Relation: flew to	Object: Haiti
Subject: Pitt	Relation: donated In	Object: May
Subject: Pitt	Relation: received in	Object: September
Subject: Pitt	Relation: met with	Object: Barack President Obama
Subject: Pitt	Relation: met	Object: Nancy Pelosi
Subject: Pitt	Relation: met in	Object: March
Subject: Pitt	Relation: established In	Object: September
Subject: Pitt	Relation: established	Object: Pitt Jolie Foundation
Subject: Pitt Jolie Foundation	Relation: donated In	Object: June
Subject: Pitt	Relation: reaffirmed In	Object: September
Subject: Pitt	Relation: announced In	Object: January
Subject: Jolie	Relation: confirmed On	Object: January
Subject: Pitt	Relation: announced in	Object: April
Subject: Jolie	Relation: filed On	Object: September
Subject: Jolie	Relation: filed from	Object: Pitt
Subject: Pitt	Relation: sued	Object: Jolie
Subject: Pitt	Relation: accompanied In	Object: July
Subject: Pitt	Relation: accompanied	Object: Jolie
Subject: Pitt	Relation: accompanied to	Object: Addis Ababa
Subject: Pitt	Relation: was On	Object: December
Subject: Jolie	Relation: gave in	Object: Swakopmund
Subject: Jolie	Relation: gave on	Object: May
Subject: Jolie	Relation: adopted On	Object: March
Subject: Pitt	Relation: adopted on	Object: February
Subject: Jolie	Relation: confirmed At	Object: Film Cannes Festival
Subject: Jolie	Relation: confirmed in	Object: May
Subject: Jolie	Relation: filed from	Object: Pitt
Subject: Jolie	Relation: filed on	Object: September
Subject: Jolie	Relation: had until	Object: May
Subject: FBI	Relation: investigated In	Object: September
Subject: FBI	Relation: investigated	Object: Pitt
Subject: Pitt	Relation: asked	Object: Jolie
Subject: Pitt	Relation: got In	Object: September
Subject: Pitt	Relation: credits	Object: Bradley Cooper

In [13]:
import wikipedia
import spacy
import random

nlp = spacy.load("en_core_web_sm")

def extract_triplets_pos(doc):
    spans = []
    i = 0
    while i < len(doc):
        if doc[i].pos_ == "PROPN":
            start = i
            while i < len(doc) and doc[i].pos_ == "PROPN":
                i += 1
            end = i
            spans.append((doc[start:end].text, start, end))
        else:
            i += 1

    results = []
    for idx in range(len(spans) - 1):
        subj_span, subj_start, subj_end = spans[idx]
        obj_span, obj_start, obj_end = spans[idx + 1]
        in_between = doc[subj_end:obj_start]

        if any(t.pos_ == "PUNCT" for t in in_between):
            continue
        if not any(t.pos_ == "VERB" for t in in_between):
            continue
        rel_tokens = [t.text for t in in_between if t.pos_ in ["VERB", "ADP"]]
        rel_text = " ".join(rel_tokens).strip()
        if rel_text:
            results.append((subj_span, rel_text, obj_span))
    return results

def extract_triplets_dep(doc):
    heads = [t for t in doc if t.pos_ == "PROPN" and t.dep_ != "compound"]

    pn_map = {}
    for h in heads:
        pn_list = [h.text]
        for child in h.children:
            if child.dep_ == "compound" and child.pos_ == "PROPN":
                pn_list.insert(0, child.text)
        pn_map[h] = " ".join(pn_list)

    def cond_1(h1, h2):
        return (
            h1.head == h2.head
            and h1.dep_ == "nsubj"
            and h2.dep_ == "dobj"
        )

    def cond_2(h1, h2):
        return (
            h1.dep_ == "nsubj"
            and h2.dep_ == "pobj"
            and h2.head.dep_ == "prep"
            and h2.head.head == h1.head
        )

    results = []
    for i in range(len(heads)):
        for j in range(len(heads)):
            if i == j:
                continue
            h1 = heads[i]
            h2 = heads[j]
            if cond_1(h1, h2):
                verb = h1.head.text
                s_text = pn_map[h1]
                o_text = pn_map[h2]
                results.append((s_text, verb, o_text))
            elif cond_2(h1, h2):
                main_verb = h1.head.text
                prep = h2.head.text
                r_text = main_verb + " " + prep
                s_text = pn_map[h1]
                o_text = pn_map[h2]
                results.append((s_text, r_text, o_text))
    return results

evaluation_pages = [
    "Donald Trump",
    "Ruth Bader Ginsburg",
    "J. K. Rowling"
]

print("Q3: Part A - Report total triplets from both extractors")
for page_title in evaluation_pages:
    content = wikipedia.page(page_title, auto_suggest=False).content
    doc_eval = nlp(content)

    q1_trips = extract_triplets_pos(doc_eval)
    q2_trips = extract_triplets_dep(doc_eval)

    print(f"\nPage: {page_title}")
    print(f"  Q1 (POS-based) triplets: {len(q1_trips)}")
    print(f"  Q2 (Dependency-based) triplets: {len(q2_trips)}")

print("Q3: Part B - Randomly Sample 5 Triplets per Page (Manual Check)")

def sample_triplets_by_page(pages, extract_fn, extractor_label):
    print(f"\n--- {extractor_label} Random Samples ---")
    for ptitle in pages:
        docp = nlp(wikipedia.page(ptitle, auto_suggest=False).content)
        all_trips = extract_fn(docp)
        sample_size = min(5, len(all_trips))
        sampled = random.sample(all_trips, sample_size) if sample_size > 0 else []

        print(f"\nPage: {ptitle} - Showing {sample_size} random triplets:")
        for s, r, o in sampled:
            print(f"  (\"{s}\", \"{r}\", \"{o}\")")

sample_triplets_by_page(evaluation_pages, extract_triplets_pos, "Q1 (POS-based)")

sample_triplets_by_page(evaluation_pages, extract_triplets_dep, "Q2 (Dependency-based)")

=== Q3: Part A - Report total triplets from both extractors ===

Page: Donald Trump
  Q1 (POS-based) triplets: 238
  Q2 (Dependency-based) triplets: 110

Page: Ruth Bader Ginsburg
  Q1 (POS-based) triplets: 171
  Q2 (Dependency-based) triplets: 75

Page: J. K. Rowling
  Q1 (POS-based) triplets: 122
  Q2 (Dependency-based) triplets: 44

=== Q3: Part B - Randomly Sample 5 Triplets per Page (Manual Check) ===

--- Q1 (POS-based) Random Samples ---

Page: Donald Trump - Showing 5 random triplets:
  ("September", "go into on", "December")
  ("Trump", "announced for of", "Europe")
  ("Justice Department", "began sent", "Trump")
  ("Taliban", "continuing on integrating", "Al")
  ("Trump", "spoke to", "Recep Tayyip Erdoğan")

Page: Ruth Bader Ginsburg - Showing 5 random triplets:
  ("Lego", "following in", "February")
  ("Ginsburg", "dictated to", "Clara Spera")
  ("Current Affairs", "criticized", "Ginsburg")
  ("Dwight D. Opperman Foundation", "established", "Ruth Bader Ginsburg Woman")
  ("P

Q3: Part A - Report total triplets from both extractors

Page: Donald Trump
  Q1 (POS-based) triplets: 238
  Q2 (Dependency-based) triplets: 110

Page: Ruth Bader Ginsburg
  Q1 (POS-based) triplets: 171
  Q2 (Dependency-based) triplets: 75

Page: J. K. Rowling
  Q1 (POS-based) triplets: 122
  Q2 (Dependency-based) triplets: 44

Q3: Part B - Randomly Sample 5 Triplets per Page (Manual Check)

--- Q1 (POS-based) Random Samples ---

Page: Donald Trump - Showing 5 random triplets:
  ("September", "go into on", "December")  INVALID
  ("Trump", "announced for of", "Europe")  INVALID
  ("Justice Department", "began sent", "Trump")  INVALID
  ("Taliban", "continuing on integrating", "Al")  VALID
  ("Trump", "spoke to", "Recep Tayyip Erdoğan")  VALID

Page: Ruth Bader Ginsburg - Showing 5 random triplets:
  ("Lego", "following in", "February")  VALID
  ("Ginsburg", "dictated to", "Clara Spera")  VALID
  ("Current Affairs", "criticized", "Ginsburg")  VALID
  ("Dwight D. Opperman Foundation", "established", "Ruth Bader Ginsburg Woman")   VALID
  ("Potawatomi", "nullify", "Prairie Band Potawatomi Nation")  VALID

Page: J. K. Rowling - Showing 5 random triplets:
  ("Scottish Arts Council", "support before", "Philosopher") INVALID
  ("Christopher Little Literary Agency", "followed", "Neil Blair")  VALID
  ("=", "= = = =", "= Works") INVALID
  ("Joanne Rowling", "born on", "July") VALID
  ("Harry Potter", "have of believe", "Harry Potter") INVALID

--- Q2 (Dependency-based) Random Samples ---

Page: Donald Trump - Showing 5 random triplets:
  ("Nancy Speaker Pelosi", "initiated in", "September") VALID
  ("TMTG", "merged In", "March")  VALID
  ("Trump", "purchased", "Lines Shuttle")  VALID
  ("Cohn", "introduced", "Roger Stone")  VALID
  ("Rosenstein", "appointed In", "May")  VALID

Page: Ruth Bader Ginsburg - Showing 5 random triplets:
  ("Ginsburg Bader Ruth Hall", "opened In", "August")  VALID
  ("Ginsburg", "wrote On", "January")  VALID
  ("Day Sandra O'Connor", "Went to", "Supreme Court")  VALID
  ("Felix Justice Frankfurter", "rejected", "Ginsburg")  VALID
  ("Deadpool", "considers for", "- Force")  INVALID

Page: J. K. Rowling - Showing 5 random triplets:
  ("Rowling", "established", "Charitable Volant Trust")  VALID
  ("Bloomsbury", "published On", "June")  VALID
  ("Rowling", "tweeted In", "September")  VALID
  ("Rowling", "shared to", "Cinema")  VALID
  ("Harry", "escapes", "Dursleys")  VALID
